src/BenchmarkDotNet/Running/BenchmarkRunnerDirty.cs

```diff
@@ -73,9 +84,86 @@ public static Summary RunSource(string source, IConfig config = null)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private static Summary RunWithDirtyAssemblyResolveHelper(Type type, IConfig config)
-            => BenchmarkRunnerClean.Run(new[] { BenchmarkConverter.TypeToBenchmarks(type, config) }).Single();
+        private static Summary RunWithDirtyAssemblyResolveHelper(Type type, IConfig config, string[] args)
+        {
+            var logger = config.GetNonNullCompositeLogger();
+            var userInteraction = new UserInteraction();
+            var (isParsingSuccess, parsedConfig, options) = ConfigParser.Parse(args, logger, config);
+            if (!isParsingSuccess) // invalid console args, the ConfigParser printed the error
+                return null;
+
+            if (options.PrintInformation)
+            {
+                logger.WriteLine(HostEnvironmentInfo.GetInformation());
+                return null;
+            }
+
+            var effectiveConfig = ManualConfig.Union(config, parsedConfig);
+
+            var types = new List<Type>();
+            types.Add(type);
+            var (allTypesValid, allAvailableTypesWithRunnableBenchmarks) = TypeFilter.GetTypesWithRunnableBenchmarks(types, new List<Assembly>(), logger);
+            if (!allTypesValid) // there were some invalid and TypeFilter printed errors
+                return null;
+
+            if (allAvailableTypesWithRunnableBenchmarks.IsEmpty())
+            {
+                userInteraction.PrintNoBenchmarksError(logger);
+                return null;
+            }
+
+            if (options.ListBenchmarkCaseMode != ListBenchmarkCaseMode.Disabled)
+            {
+                BenchmarkCasesPrinter.PrintList(logger, effectiveConfig, allAvailableTypesWithRunnableBenchmarks, options);
+                return null;
+            }
+
+            var benchmarks = options.UserProvidedFilters
+                           ? TypeFilter.Filter(effectiveConfig, allAvailableTypesWithRunnableBenchmarks)
+                           : new BenchmarkRunInfo[] { BenchmarkConverter.TypeToBenchmarks(type, config) };
 
+            return BenchmarkRunnerClean.Run(benchmarks).Single();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static Summary RunWithDirtyAssemblyResolveHelper(Assembly assembly, IConfig config, string[] args)
```

Comment:
I think that we should remove the code duplication.

We could have a single method that would handle both `Type` and `Assembly` type arguments:

```cs
private static Summary RunWithDirtyAssemblyResolveHelper(Type type, IConfig config, string[] args)
 => RunWithDirtyAssemblyResolveHelper(new [] { type }, Array.Empty<Assembly>(), config, args);

private static Summary RunWithDirtyAssemblyResolveHelper(Assembly assembly, IConfig config, string[] args)
 => RunWithDirtyAssemblyResolveHelper(Array.Empty<Type>(), new [] { assembly }, config, args);

private static Summary RunWithDirtyAssemblyResolveHelper(IEnumerable<Type> types, IEnumerable<Assembly> assemblies, IConfig config, string[] args)
{
  // skipped for brevity
  var (allTypesValid, allAvailableTypesWithRunnableBenchmarks) = TypeFilter.GetTypesWithRunnableBenchmarks(types, assemblies, logger);
}
```

Reply:
why would, IEnumerable<Type> types, IEnumerable<Assembly> assemblies will be IEnumerable? when run method is getting passed only one item of type and assembly! Is any other reason? 

Other way i think it might be better! passing ``` assembly.GetRunnableBenchmarks().ToArray() ``` to ``` RunWithDirtyAssemblyResolveHelper(Type type, IConfig config, string[] args) ``` with IEnumerable<Type> types..

by passing assembly types, we dont need to write ``` RunWithDirtyAssemblyResolveHelper() ``` that takes ```types, config and args[] and assembly``` as parameters and using some if validation to apply type rules to run benchmark.

Reply:
> why would, IEnumerable types, IEnumerable assemblies will be IEnumerable? when run method is getting passed only one item of type and assembly! Is any other reason?

You are right, we don't need an `IEnumerable` here.

> Other way i think it might be better!

The assemblies have a "special" treatment because they might also contain a generic benchmark types and unfortunately this would break the generic type benchmarks .

https://github.com/dotnet/BenchmarkDotNet/blob/2415fda6fb8eaf2d0f17b1828a7ee9befbd7baab/samples/BenchmarkDotNet.Samples/IntroGenericTypeArguments.cs#L6-L8

https://github.com/dotnet/BenchmarkDotNet/blob/2ba30330ec5ec8a9abe284c927277a6fe0fc5a0c/src/BenchmarkDotNet/Running/TypeFilter.cs#L32-L35

Reply:
also, ``` TypeToBenchmarks(Type type, IConfig config = null) ``` , not ``` IEnumerable<Type> types, IEnumerable<Assembly> assemblies```